build.sbt

```diff
@@ -30,6 +30,8 @@ libraryDependencies ++= {
     "org.monifu"                  %% "monifu"                               % "1.0-RC3",
     "org.slf4j"                    % "slf4j-api"                            % "1.7.7",
     "ch.qos.logback"               % "logback-classic"                      % "1.0.13",
+    "org.scalaz"                  %% "scalaz-core"                          % "7.2.1",
```

Comment:
I did consider this option.

The first implementation was a simple mutable Map. 

It would either be 

```
val state = mutable.Map[String, Filter]
```

or

```
// the var !!! and not val.
var state = immutable.Map[String, Filter]
```

Everywhere we use `val` and `Immutable`. So let's go immutable, and there should be a way to manage state simple without complications. And we shouldn't be dealing with too many intermediate state variables which become stale the next line. So the State Monad with all the function composition is one good way to deal with it. `Scalaz` provides amazing helpers to deal with it.

We could add all the types and helpers we use from Scalaz. It's just `State` monad with `modify`, `gets` and `state constructor` methods and the `run` function. But if it's already there in scalaz, IMO, it would just be to remove one dependency that we should do this.

I'm not sure if this is the right way to do this. I'm happy to rewrite it with simple constructs, share your ideas.